src/main/java/graphql/util/LazyReference.java

```diff
@@ -0,0 +1,57 @@
+package graphql.util;
+
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A lazy value that is computed once and once only.  Useful for effectively final values
+ * that dont need to be computed on creation but on first time use.
+ *
+ * @param <T> for two
+ */
+public final class LazyReference<T> {
+
+    private final Supplier<T> valueSupplier;
+    private volatile T value;
+
+    public LazyReference(Supplier<T> valueSupplier) {
+        this.valueSupplier = requireNonNull(valueSupplier);
+    }
+
+    /**
+     * @return the computed lazy value
+     */
+    public T get() {
```

Comment:
this should follow the https://en.wikipedia.org/wiki/Double-checked_locking for Java 5:
```java
class Foo {
    private volatile Helper helper;
    public Helper getHelper() {
        Helper localRef = helper;
        if (localRef == null) {
            synchronized (this) {
                localRef = helper;
                if (localRef == null) {
                    helper = localRef = new Helper();
                }
            }
        }
        return localRef;
    }

    // other functions and members...
}````